import { faker } from '@faker-js/faker';

import { SignUpController } from '@/presentation/controllers';
import {
  ServerError,
  EmailInUseError,
  InvalidParamError,
  MissingParamError,
} from '@/presentation/errors';
import { badRequestError, forbiddenError, ok, serverError } from '@/presentation/helpers';
import { AddAccountSpy, ValidationSpy, AuthenticationSpy } from '@/tests/presentation/mocks';

const mockRequest = (): SignUpController.Request => {
  const password = faker.internet.password();
  return {
    name: faker.name.fullName(),
    email: faker.internet.email(),
    password,
    passwordConfirmation: password,
  };
};

interface SutTypes {
  sut: SignUpController;
  validationSpy: ValidationSpy;
  addAccountSpy: AddAccountSpy;
  authenticationSpy: AuthenticationSpy;
}

const makeSut = (): SutTypes => {
  const addAccountSpy = new AddAccountSpy();
  const validationSpy = new ValidationSpy();
  const authenticationSpy = new AuthenticationSpy();
  const sut = new SignUpController(validationSpy, addAccountSpy, authenticationSpy);

  return {
    sut,
    validationSpy,
    addAccountSpy,
    authenticationSpy,
  };
};

describe('SignUp Controller', () => {
  test('Should throw if Validation throws', async () => {
    const { sut, validationSpy } = makeSut();
    validationSpy.error = new InvalidParamError('email');
    const response = await sut.handle(mockRequest());
    expect(response).toEqual(badRequestError(validationSpy.error));
  });

  test('Should call Validation correctly', async () => {
    const { sut, validationSpy } = makeSut();
    const validateSpy = jest.spyOn(validationSpy, 'validate');
    const request = mockRequest();
    await sut.handle(request);
    expect(validateSpy).toHaveBeenCalledWith(request);
  });

  test('Should return 400 if Validation returns an error', async () => {
    const { sut, validationSpy } = makeSut();
    validationSpy.error = new MissingParamError(faker.random.word());
    const httpResponse = await sut.handle(mockRequest());
    expect(httpResponse).toEqual(badRequestError(validationSpy.error));
  });

  test('Should throw if AddAccount throws', async () => {
    const { sut, addAccountSpy } = makeSut();
    jest.spyOn(addAccountSpy, 'add').mockImplementationOnce(async () => {
      return await Promise.reject(new ServerError(null));
    });
    const response = await sut.handle(mockRequest());
    expect(response).toEqual(serverError(new ServerError(null)));
  });

  test('Should call AddAccount with correct params', async () => {
    const { sut, addAccountSpy } = makeSut();
    const addSpy = jest.spyOn(addAccountSpy, 'add');
    await sut.handle(mockRequest());
    expect(addSpy).toHaveBeenCalledWith(addAccountSpy.params);
  });

  test('Should return 403 when account exists', async () => {
    const { sut, addAccountSpy } = makeSut();
    addAccountSpy.result = false;
    const response = await sut.handle(mockRequest());
    expect(response).toEqual(forbiddenError(new EmailInUseError()));
  });

  test('Should throw if Authentication throws', async () => {
    const { sut, authenticationSpy } = makeSut();
    jest.spyOn(authenticationSpy, 'auth').mockImplementationOnce(async () => {
      return await Promise.reject(new ServerError(null));
    });
    const response = await sut.handle(mockRequest());
    expect(response).toEqual(serverError(new ServerError(null)));
  });

  test('Should call Authentication with correct params', async () => {
    const { sut, authenticationSpy } = makeSut();
    const authSpy = jest.spyOn(authenticationSpy, 'auth');
    const request = mockRequest();
    await sut.handle(request);
    expect(authSpy).toHaveBeenCalledWith(request);
  });

  test('Should return 200 if ok', async () => {
    const { sut, authenticationSpy } = makeSut();
    const response = await sut.handle(mockRequest());
    expect(response).toEqual(ok(authenticationSpy.result));
  });
});
